### PR TITLE
Run translation and LLVM in parallel when compiling with multiple CGUs

### PR DESCRIPTION
**Pull request by michaelwoerister**
This is still a work in progress but the bulk of the implementation is done, so I thought it would be good to get it in front of more eyes.

This PR makes the compiler start running LLVM while translation is still in progress, effectively allowing for more parallelism towards the end of the compilation pipeline. It also allows the main thread to switch between either translation or running LLVM, which allows to reduce peak memory usage since not all LLVM module have to be kept in memory until linking. This is especially good for incr. comp. but it works just as well when running with `-Ccodegen-units=N`.

In order to help tuning and debugging the work scheduler, the PR adds the `-Ztrans-time-graph` flag which spits out html files that show how work packages where scheduled:
Building regex
(red is translation, green is llvm)

One side effect here is that `-Ztime-passes` might show something not quite correct because trans and LLVM are not strictly separated anymore. I plan to have some special handling there that will try to produce useful output.

One open question is how to determine whether the trans-thread should switch to intermediate LLVM processing.

TODO:
- [x] Restore `-Z time-passes` output for LLVM.
- [x] Update documentation, esp. for work package scheduling.
- [x] Tune the scheduling algorithm.

cc @alexcrichton @rust-lang/compiler

### CODE REVIEW

**Comment by michaelwoerister** [comment]
r? @alexcrichton 

Pre-assigning @alexcrichton for review, so he can already start reading `:P`

**Comment by alexcrichton** [comment]
🎊

**Comment by retep998** [comment]
As a future thought for how `-Ztime-passes` could be implemented in a world where everything is threaded, we could use `GetThreadTimes` on Windows (and whatever linux equivalent) to measure the CPU time used up by a given thread to more precisely track where time is being spent.

File: src/librustc_trans/back/write.rs [hunk 1]
```diff
@@ -1109,164 +1201,230 @@ fn execute_work<'a>(sess: &'a Session,
...
+                        let cgcx = CodegenContext {
+                            worker: TRANS_WORKER_ID,
+                            .. cgcx.clone()
+                        };
+                        trans_worker_state = TransWorkerState::LLVMing;
+                        spawn_work(cgcx, item);
+                    }
+                }
+            } else {
+                match trans_worker_state {
```
Comment (alexcrichton) [inline src/librustc_trans/back/write.rs]: Ok so to see if I understand this:

* The documentation above refers to a "translation worker"
* This trans worker seems to represent the "ephemeral token" that we inherently have to run work from a jobserver
* The literal thread representing the trans worker can change I think? Sometimes it's the literally the thread doing translation, sometimes it's a spawned worker here to translate an existing module.
* If we happen to reach 4 translated codegen units but not codegen'd codegen units then we request the translating main thread to stop, and continue its work with a freshly spawned thread to codegen a module.

Does that sound roughly right?

Ah I see now there's also a crucial piece where when *any* codegen thread finishes we consider it "trans worker" thread now available again. This means after any codegen thread finishes we may be candidate to start translation of another unit of work I think, right?

**Review by alexcrichton — COMMENTED** [review]
Looking great!

The general framework of who's running what when seemed a little confusing to follow, but I think a comment would go a long way towards helping that.

File: src/librustc_trans/back/write.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: > Ah I see now there's also a crucial piece where when any codegen thread finishes we consider it "trans worker" thread now available again. This means after any codegen thread finishes we may be candidate to start translation of another unit of work I think, right?

No, but it's an excellent idea `:D`

File: src/librustc_trans/back/write.rs [hunk 2]
```diff
@@ -1109,164 +1201,230 @@ fn execute_work<'a>(sess: &'a Session,
...
+
+            match coordinator_receive.recv().unwrap() {
+                // Save the token locally and the next turn of the loop will use
+                // this to spawn a new unit of work, or it may get dropped
+                // immediately if we have no more work to spawn.
+                Message::Token(token) => {
+                    if let Ok(token) = token {
+                        tokens.push(token);
+                    } else {
+                        shared_emitter.fatal("failed to acquire jobserver token");
```
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: What do you mean? Just panic?

File: src/librustc_trans/back/write.rs [hunk 3]
```diff
@@ -1109,164 +1201,230 @@ fn execute_work<'a>(sess: &'a Session,
...
+                        panic!()
                     }
-                    None => {
-                        handler.emit(&MultiSpan::new(),
-                                     &diag.msg,
-                                     diag.lvl);
+                }
+
+                Message::TranslationDone { llvm_work_item, is_last } => {
+                    work_items.insert(0, llvm_work_item);
```
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: No reason. I changed that during a debugging session. I'll switch it back to a `push`.

File: src/librustc_trans/back/write.rs [hunk 4]
```diff
@@ -1109,164 +1201,230 @@ fn execute_work<'a>(sess: &'a Session,
...
+                // "ephemeral token"
+                //
+                // Note that if the thread failed that means it panicked, so we
+                // abort immediately.
+                Message::Done { result: Ok(compiled_module), worker_id } => {
+                    if worker_id == TRANS_WORKER_ID {
+                        assert_eq!(trans_worker_state, TransWorkerState::LLVMing);
+                        trans_worker_state = TransWorkerState::Idle;
+                    } else {
+                        drop(tokens.pop());
```
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: Yes, good idea. The `truncate` above should take care of dropping the `Token` if we actually don't need it.

File: src/librustc_trans/back/write.rs (hunk 2 above)
Comment (alexcrichton) [inline src/librustc_trans/back/write.rs]: Oh just something like:

```rust
match token {
    Ok(token) => tokens.push(token),
    Err(e) => shared_emitter.fatal(&format!("failed to acquire jobserver token: {}", e)),
}
```

**Comment by alexcrichton** [comment]
Docs look great! So one overall meta-comment as well now. I'm having some difficulty articulating this but I'm slightly worried about a situation like:

* We've got N codegen units left to translate, but the main thread is stopped
* Our heuristic blocks the main thread an spins up a worker for an existing codegen unit.
* All of a sudden we get an influx of jobserver tokens, but the main thread is still stopped.

I haven't convinced myself this is possible and I think that the various handling makes it ok? I've found the interaction between all these threads and the implicit token a little difficult to follow, but does this make sense to you? Can you think of a case where we accidentally starve the translation thread due to our heuristic?

**Comment by michaelwoerister** [comment]
> I haven't convinced myself this is possible and I think that the various handling makes it ok? I've found the interaction between all these threads and the implicit token a little difficult to follow, but does this make sense to you? Can you think of a case where we accidentally starve the translation thread due to our heuristic?

So with the latest change, the main thread will be free again as soon as the first LLVM worker is done (thanks to your suggestion). It's still possible to run into a LLVM work shortage the way you describe though.

I just adapted the strategy to estimate the cost of a LLVM WorkItem. Maybe the main thread should always start the cheapest one available, so it can get back to translating sooner, if needed? I don't think this is much of a problem though (with no data to back this claim up in any way `:)`)

File: src/librustc_trans/back/write.rs (hunk 2 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/write.rs]: ah ok, yes.

**Comment by alexcrichton** [comment]
Thinking about this a bit more, I think this is my mental model for what's happening: on each turn of the loop we'll have N slots of work to fill up depending on what's currently running and what amount of tokens we have. Given the choice of whether to translate a new unit or codegen an existing unit it seems fine to have a heuristic. Whenever something happens though it'll turn the loop and cause everything to start over. 

I think that's roughly what's implemented right now, but I think that it means that we should consider the translation thread idle as soon as we've acquired a new token? That way if the translation thread was blocked and we get a token it should get unblocked? (which I don't think happens today?) I may not be following the code quite right though...

> I just adapted the strategy to estimate the cost of a LLVM WorkItem

Neat!

**Comment by michaelwoerister** [comment]
> but I think that it means that we should consider the translation thread idle as soon as we've acquired a new token? 

Yes, that makes sense. It's a variation of considering the translation thread idle when a package is finished (another way of getting an additional token).

**Comment by michaelwoerister** [comment]
@alexcrichton I think your descriptions of the system are accurate. Also, the current heuristic will never let the translation thread do LLVM work in "normal" multi-CGU scenarios. Only when the number of CGUs vastly outnumbers the number of CPU cores (as in an incr. comp. scenario) will the different scheduling kick in. So we should not see any throughput regressions here.

(As a side note: in memory constrained scenarios, it might make sense to use a different heuristic. But that's something for another time, I guess).

I've restored `-Ztime-passes` output for trans and LLVM now -- in a way that seems reasonable to me. For single-CGU sessions it should be equivalent to what we have now. For multi-CGU sessions, it will start recording the `LLVM passes` time when the first regular LLVM module is received.

I think all comments are addressed now.

**Comment by kennytm** [comment]
`compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.rs` failed to fail.

<details>

```
[00:49:15] ---- [compile-fail] compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.rs stdout ----
[00:49:15] 	
[00:49:15] error: /checkout/src/test/compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.rs:876: expected error not found: compilation successful
[00:49:15] 
[00:49:15] error: 0 unexpected errors found, 1 expected errors not found
[00:49:15] status: exit code: 101
[00:49:15] command: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc /checkout/src/test/compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.rs -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail --target=x86_64-unknown-linux-gnu --error-format json -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.stage2-x86_64-unknown-linux-gnu.compile-fail.libaux -C prefer-dynamic -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.stage2-x86_64-unknown-linux-gnu -Crpath -O -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers
[00:49:15] not found errors (from test file): [
[00:49:15]     Error {
[00:49:15]         line_num: 876,
[00:49:15]         kind: Some(
[00:49:15]             Error
[00:49:15]         ),
[00:49:15]         msg: "compilation successful"
[00:49:15]     }
[00:49:15] ]
[00:49:15] 
[00:49:15] thread '[compile-fail] compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:1133:12
[00:49:15] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:49:15] 
[00:49:15] 
[00:49:15] failures:
[00:49:15]     [compile-fail] compile-fail/feature-gate/issue-43106-gating-of-builtin-attrs.rs
[00:49:15] 
[00:49:15] test result: FAILED. 2703 passed; 1 failed; 13 ignored; 0 measured; 0 filtered out
```

</details>

Line 876 of the test:

```rust
// Since we expect for the mix of attributes used here to compile
// successfully, and we are just testing for the expected warnings of
// various (mis)uses of attributes, we use the `rustc_error` attribute
// on the `fn main()`.

#[rustc_error]
fn main() { //~ ERROR compilation successful
    println!("Hello World");
}
```

Seems like `#[rustc_error]` is ignored or the message is no longer produced. The message is produced at `src/librustc_trans/base.rs:665`.

**Comment by michaelwoerister** [comment]
Interesting. I can reproduce this locally but I don't know what the root cause is. The `#[rustc_error]` attribute is used in dozens of other tests with problem.

**Comment by michaelwoerister** [comment]
I see, it fails with a different, valid error that was not reached before the refactoring. I'll move the check for `#[rustc_error]` to an earlier position.

**Comment by kennytm** [comment]
Many incremental tests failed due to `unwrap()`s. Legit.

```
[00:47:20] failures:
[00:47:20]     [incremental] incremental/change_add_field/struct_point.rs
[00:47:20]     [incremental] incremental/change_private_fn/struct_point.rs
[00:47:20]     [incremental] incremental/change_private_impl_method/struct_point.rs
[00:47:20]     [incremental] incremental/change_pub_inherent_method_body/struct_point.rs
[00:47:20]     [incremental] incremental/change_pub_inherent_method_sig/struct_point.rs
[00:47:20]     [incremental] incremental/foreign.rs
[00:47:20]     [incremental] incremental/inlined_hir_34991/main.rs
[00:47:20]     [incremental] incremental/issue-42602.rs
[00:47:20]     [incremental] incremental/krate-inlined.rs
[00:47:20]     [incremental] incremental/remove-private-item-cross-crate/main.rs
```

**Comment by michaelwoerister** [comment]
@bors r=alexcrichton

I could not reproduce this locally. I added more error output. Let's see if that gives a clue about what's wrong here. I'm wondering though why RUST_BACKTRACE doesn't seem to work...

**Comment by kennytm** [comment]
@michaelwoerister Travis CI failed again with the same errors. Maybe an LLVM-3.7 issue?

<details>

```
...

[00:45:55] ---- [incremental] incremental/remove-private-item-cross-crate/main.rs stdout ----
[00:45:55] 	
[00:45:55] error in revision `rpass2`: compilation failed!
[00:45:55] status: exit code: 101
[00:45:55] command: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc /checkout/src/test/incremental/remove-private-item-cross-crate/main.rs -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental --target=x86_64-unknown-linux-gnu --cfg rpass2 -Z incremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/remove-private-item-cross-crate/main.inc --error-format json -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/remove-private-item-cross-crate/main.stage2-x86_64-unknown-linux-gnu.incremental.libaux -C prefer-dynamic -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental/remove-private-item-cross-crate/main.stage2-x86_64-unknown-linux-gnu -Crpath -O -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers -Zincremental-info
[00:45:55] stdout:
[00:45:55] ------------------------------------------
[00:45:55] 
[00:45:55] ------------------------------------------
[00:45:55] stderr:
[00:45:55] ------------------------------------------
[00:45:55] thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
[00:45:55] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:45:55] {"message":"Error during translation/LLVM phase.","code":null,"level":"error","spans":[],"children":[],"rendered":null}
[00:45:55] 
[00:45:55] ------------------------------------------
[00:45:55] 
[00:45:55] thread '[incremental] incremental/remove-private-item-cross-crate/main.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2499:8
[00:45:55] 
[00:45:55] 
[00:45:55] failures:
[00:45:55]     [incremental] incremental/change_add_field/struct_point.rs
[00:45:55]     [incremental] incremental/change_private_fn/struct_point.rs
[00:45:55]     [incremental] incremental/change_private_impl_method/struct_point.rs
[00:45:55]     [incremental] incremental/change_pub_inherent_method_body/struct_point.rs
[00:45:55]     [incremental] incremental/change_pub_inherent_method_sig/struct_point.rs
[00:45:55]     [incremental] incremental/foreign.rs
[00:45:55]     [incremental] incremental/inlined_hir_34991/main.rs
[00:45:55]     [incremental] incremental/krate-inlined.rs
[00:45:55]     [incremental] incremental/remove-private-item-cross-crate/main.rs
[00:45:55] 
[00:45:55] test result: FAILED. 69 passed; 9 failed; 0 ignored; 0 measured; 0 filtered out
```

</details>

**Comment by michaelwoerister** [comment]
Thanks, @kennytm. Compiling with LLVM 3.7 right now...

**Comment by michaelwoerister** [comment]
Can't reproduce with LLVM 3.7 either `:(`

**Comment by michaelwoerister** [comment]
I pushed another change that should give a sensible error message at a likely point of failure. Also adapted the scheduler heuristic slightly. Let's see.

**Comment by michaelwoerister** [comment]
New error, excellent `:D`

**Comment by michaelwoerister** [comment]
@bors r=alexcrichton

OK, passes travis now.

**Pull request merged**